icefoganalytics/travel-authorization
# Make Departure Times Optional In Travel Authorization Form And Various Other Locations

Fixes https://github.com/icefoganalytics/travel-authorization/issues/120

Relates to:

- https://github.com/icefoganalytics/travel-authorization/issues/122

# Context

**Is your feature request related to a problem? Please describe.**

In many cases travelers will not know their exact departure times. If they know it or put something in great. If not we need to assume they are travelling all day for the estimates.

**Describe the solution you'd like**

1. Make travel authorization -> stops departure times optional at the form level.
2. Make travel authorization -> stops departure times optional at the database level.
3. Make travel authorization -> stops departure times optional in the estimate generator.
    1. Assume all day travel if time is not supplied, i.e. maximize cost estimate per-diems

## Concerns

This way of building the form with travel time optional, seems like it would encourage people to avoid tracking their departure times. In the long run it would likely be better to simply make departure time tracking easy enough that requiring the user to enter it isn't a problem for them.

# Implementation

1. Add `TravelSegment` model and associated table to replace Stop model in the future.
2. Upgrade estimate generation so that it relies on travel segments instead of stops.
3. Fix a bunch of bugs in estimate generation.
4. Make travel authorization update create travel segments from stops. This means that you must run an update before trying to generate an estimate. The current UI is doing that, but any external API's need to be aware of this limitation.
5. Built a `BulkConvertStopsToTravelSegmentsService` in the "stops" namespace so that it's easy to remove the the code later. This service handles all the weirdness around converting stops to travel segments.
6. Stripped out all `id` generation from all factory `build` actions, this avoid creating id conflicts in the database.
7. Switch database cleanup to raw truncates of all tables.

# Screenshots

![image](https://github.com/icefoganalytics/travel-authorization/assets/23045206/ea2ed657-f559-4226-bdbc-48ea470dc806)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Go to the "My Travel Request" page via the top drop down nav.
5. Create a new travel request and fill in the form.
6. Check that you can generate estimates.
